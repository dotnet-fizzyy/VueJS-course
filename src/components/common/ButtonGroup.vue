<template>
    <div :class="$style.root">
        <span :class="$style.label" v-if="!!label">{{ label }}</span>

        <div :class="$style['options-container']">
            <template v-for="(option, index) in options">
                <button
                    :class="[
                        isSelected(option.name) ? $style['selected-option'] : $style['non-selected-option'],
                        $style.option,
                    ]"
                    :key="index"
                    @click="selectOption(option.name)"
                >
                    {{ option.label }}
                </button>
            </template>
        </div>
    </div>
</template>

<script lang="ts">
import Vue from 'vue';

export interface ButtonGroupOption {
    label: string;
    name: string;
}

export interface ButtonGroupProps {
    label: string;
    options: ButtonGroupOption[];
    selectedOption: string;
    selectOption: (value: string) => void;
}

export default Vue.extend({
    name: 'ButtonGroup',
    props: {
        label: {
            type: String,
        },
        options: {
            type: Array as Vue.PropType<ButtonGroupOption[]>,
            required: true,
        },
        selectedOption: {
            type: String,
            required: true,
        },
    },
    methods: {
        selectOption(option: string): void {
            if (option === this.selectedOption) {
                return;
            }

            this.$emit('select-option', option);
        },
        isSelected(optionName: string): boolean {
            return optionName === this.selectedOption;
        },
    },
});
</script>

<style lang="scss" scoped module>
@import '../../styles/buttons';
@import '../../styles/fonts';

.root {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: center;
}

.label {
    @extend .font;

    text-transform: uppercase;
    margin-right: 20px;
}

.options-container {
    display: flex;
    flex-direction: row;
    height: 35px;
}

.selected-option {
    @extend .primary-button;
}

.non-selected-option {
    @extend .secondary-button;
}

.option {
    &:first-child {
        border-top-right-radius: 0;
        border-bottom-right-radius: 0;
    }

    &:last-child {
        border-top-left-radius: 0;
        border-bottom-left-radius: 0;
    }
}
</style>
