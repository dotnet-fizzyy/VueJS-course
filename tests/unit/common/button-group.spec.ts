import ButtonGroup from '@/components/common/ButtonGroup.vue';
import { shallowMount } from '@vue/test-utils';

describe('ButtonGroup.vue Tests', () => {
    it('Should render', () => {
        //Arrange
        const selectedOption = 'btn2';

        const wrapper = shallowMount(ButtonGroup, {
            propsData: {
                label: 'Label',
                options: [
                    { label: 'First Button', name: 'btn1' },
                    { label: 'Second Button', name: selectedOption },
                ],
                selectedOption,
            },
        });

        //Act & Assert
        expect(wrapper).toMatchSnapshot();
    });

    it('Should not render "label" block if it was not provided via props', () => {
        //Arrange
        const selectedOption = 'btn2';

        const wrapper = shallowMount(ButtonGroup, {
            propsData: {
                options: [
                    { label: 'First Button', name: 'btn1' },
                    { label: 'Second Button', name: selectedOption },
                ],
                selectedOption,
            },
        });

        //Act & Assert
        expect(wrapper).toMatchSnapshot();
    });

    it('Should emit event if non-selected button was clicked', () => {
        //Arrange
        const nonSelectedOption = 'bt1';
        const selectedOption = 'btn2';

        const wrapper = shallowMount(ButtonGroup, {
            propsData: {
                label: 'Label',
                options: [
                    { label: 'First Button', name: nonSelectedOption },
                    { label: 'Second Button', name: selectedOption },
                ],
                selectedOption,
            },
        });

        //Act & Assert
        const nonSelectedButtonWrapper = wrapper.find(`#${nonSelectedOption}`);

        expect(nonSelectedButtonWrapper.element).toBeTruthy();

        nonSelectedButtonWrapper.trigger('click');

        expect(wrapper.emitted()['select-option']).toBeTruthy();
        expect(wrapper.emitted()['select-option'][0][0]).toEqual(nonSelectedOption);
    });

    it('Should not emit event if selected button was clicked', () => {
        //Arrange
        const nonSelectedOption = 'bt1';
        const selectedOption = 'btn2';

        const wrapper = shallowMount(ButtonGroup, {
            propsData: {
                label: 'Label',
                options: [
                    { label: 'First Button', name: nonSelectedOption },
                    { label: 'Second Button', name: selectedOption },
                ],
                selectedOption,
            },
        });

        //Act & Assert
        const selectedButtonWrapper = wrapper.find(`#${selectedOption}`);

        expect(selectedButtonWrapper.element).toBeTruthy();

        selectedButtonWrapper.trigger('click');

        expect(wrapper.emitted()['select-option']).toBeFalsy();
    });
});
