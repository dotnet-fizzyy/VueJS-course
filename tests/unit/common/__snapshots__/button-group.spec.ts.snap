// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonGroup.vue Tests Should not render "label" block if it was not provided via props 1`] = `
<div class="root">
  <!---->
  <div class="options-container"><button data-aqa-button-name="btn1" class="non-selected-option option">
      First Button
    </button><button data-aqa-button-name="btn2" data-aqa-is-button-selected="true" class="selected-option option">
      Second Button
    </button></div>
</div>
`;

exports[`ButtonGroup.vue Tests Should render 1`] = `
<div class="root"><span class="label">Label</span>
  <div class="options-container"><button data-aqa-button-name="btn1" class="non-selected-option option">
      First Button
    </button><button data-aqa-button-name="btn2" data-aqa-is-button-selected="true" class="selected-option option">
      Second Button
    </button></div>
</div>
`;
